Feature/url shortener tamirlan dexengazy

dextomi opened the pull request:
shortener done and it somehow works now
short link now redirects as it should

dextomi committed:
a little bit of shortener done

dextomi committed:
url shortener done

mysosnovskaya commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,146 @@
+    private final Executor hashCacheExecutor;
+    private final Executor hashGeneratorExecutor;
+
+    public HashService(
```
Конструктор должен быть после полей) Ну и можно использовать конструктор от ломбока, нужно только ломбоку рассказать про аннотации Qualifier. Это в lombok.config файле настраивается. Оставлю как задачку на подумать :)

dextomi replied:
Принял, знаю про это, привык руками писать в таких случаях конструктор, теперь так делать не буду. Наверное...

mysosnovskaya commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,146 @@
+    private int lockId;
+
+    public CompletableFuture<List<String>> getHashesAsync(long count) {
+        return CompletableFuture.supplyAsync(() -> getHashes(count), hashCacheExecutor);
```
Будет ли работать транзакция у getHashes при таком вызове?

dextomi replied:
Нет, не должна, что-то поспешил и забыл вынести асинхронные методы в отдельный класс. Как вариант можно и через @Async сделать

mysosnovskaya commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,146 @@
+        return hashList;
+    }
+
+    private void generateHashBatches() {
```
Эта логика должна быть в отдельном классе - генераторе

dextomi replied:
Да, там можно еще пару методов по хорошему в разные классы вынести, но тут решил просто все в одном классе сделать, из-за чего как раз и забыл вынести асинхронные методы отдельно

mysosnovskaya commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,146 @@
+                .toList();
+    }
+
+    private List<Hash> encodeBatch(List<Long> batchList) {
```
Этот метод можно в base62Encoder вынести

dextomi replied:
Полностью согласен

dextomi replied:
Вернулся шортенер править, подумал немного и пришел к выводу что так и должно быть. У меня энкодер просто должен энкодить что-то. А тут я именно батчи хэшей собираю и кодирую. По этой причине думаю что так и должно остаться

mysosnovskaya commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,146 @@
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+
+        Collections.shuffle(hashes);
```
Зачем шафл?

dextomi replied:
Чтобы не перемешивать все в БД и гарантировать что будут как-то рандомно хэши выдаваться.
Шафл быстрее чем через рандом в БД

mysosnovskaya commented on src/main/java/faang/school/urlshortenerservice/storage/HashMemoryCache.java:
```diff
@@ -0,0 +1,52 @@
+@RequiredArgsConstructor
+public class HashMemoryCache {
+    private final HashService hashService;
+    private final AtomicBoolean isFull = new AtomicBoolean(false);
```
Это поле точно должно так называться? Не очевидно, что это значит

dextomi replied:
Да, надо бы переименовать её

mysosnovskaya reviewed (commented):
На сертификат зачтено

dextomi committed:
Created lombok configuration in the root directory;
Moved async method to another class to make transactions work;
Renamed variable isFull -> isCacheRefilling for it to be clear
